Restoring chain with `StateRootInHeader = true` can fail

# Problem:
`db restore` results depend on the db state. The state root stored for genesis block is not valid if service was started with non-empty database containing a single genesis block.

# To reproduce
The test to reproduce (put it to `pkg/core`).
BTW, this is the easiest way I have found to share storage between chains, could you tell me if this can be done easier?
```
type nopCloserStorage struct {
	storage.Store
}

func (nopCloserStorage) Close() error {
	return nil
}

func TestBlockchainRestoreStateRootInHeader(t *testing.T) {
	bc := newTestChainWithCustomCfg(t, func(c *config.Config) {
		c.ProtocolConfiguration.StateRootInHeader = true
	})
	require.NoError(t, bc.AddBlock(bc.newBlock()))

	w := io.NewBufBinWriter()
	require.NoError(t, chaindump.Dump(bc, w.BinWriter, 0, 2))
	require.NoError(t, w.Err)

	data := w.Bytes()
	fcfg := func(c *config.Config) { c.ProtocolConfiguration.StateRootInHeader = true }
	initChain := func(st storage.Store) *Blockchain {
		chain := initTestChain(t, st, fcfg)
		go chain.Run()
		return chain
	}

	t.Run("this passes", func(t *testing.T) {
		bc := initChain(storage.NewMemoryStore())
		r := io.NewBinReaderFromBuf(data)
		require.NoError(t, chaindump.Restore(bc, r, 0, 2, nil))
		bc.Close()
	})
	t.Run("this doesn't pass", func(t *testing.T) {
		st := nopCloserStorage{Store: storage.NewMemoryStore()}
		{
			bc := initChain(st)
			r := io.NewBinReaderFromBuf(data)
			require.NoError(t, chaindump.Restore(bc, r, 0, 1, nil))
			bc.Close()
		}
		{
			bc := initChain(st)
			r := io.NewBinReaderFromBuf(data)
			require.NoError(t, chaindump.Restore(bc, r, 1, 1, nil))
			bc.Close()
		}
	})
}
```

Results:
```
$ go test -count=1 -v .
...
=== NAME  TestBlockchainRestoreStateRootInHeader/this_doesn't_pass
    blockchain_core_test.go:185:
                Error Trace:    /repo/neo-go/pkg/core/blockchain_core_test.go:185
                Error:          Received unexpected error:
                                failed to add block 1: state root for previous block is invalid: 6fc66e00ccd8ed7ae0563e9d2f2b859c2abbede741fb546ceb99e5253e8aa41d != 0000000000000000000000000000000000000000000000000000000000000000
                Test:           TestBlockchainRestoreStateRootInHeader/this_doesn't_pass
=== NAME  TestBlockchainRestoreStateRootInHeader
    logger.go:146: 2024-10-04T17:24:54.486+0300 INFO    persisted to disk       {"blocks": 1, "keys": 141, "headerHeight": 1, "blockHeight": 1, "took": "27.268µs"}
--- FAIL: TestBlockchainRestoreStateRootInHeader (0.01s)
    --- PASS: TestBlockchainRestoreStateRootInHeader/this_passes (0.00s)
    --- FAIL: TestBlockchainRestoreStateRootInHeader/this_doesn't_pass (0.00s)
FAIL
```

I have quickly glanced upon tests for chaindump and it seems we never close the blockchain in those tests, just combine multiple `Restore()` calls.

--- a/pkg/core/blockchain_core_test.go
+++ b/pkg/core/blockchain_core_test.go
@@ -12,9 +12,11 @@ import (
 	"github.com/nspcc-dev/neo-go/internal/testchain"
 	"github.com/nspcc-dev/neo-go/pkg/config"
 	"github.com/nspcc-dev/neo-go/pkg/core/block"
+	"github.com/nspcc-dev/neo-go/pkg/core/chaindump"
 	"github.com/nspcc-dev/neo-go/pkg/core/state"
 	"github.com/nspcc-dev/neo-go/pkg/core/storage"
 	"github.com/nspcc-dev/neo-go/pkg/core/transaction"
+	"github.com/nspcc-dev/neo-go/pkg/io"
 	"github.com/nspcc-dev/neo-go/pkg/smartcontract"
 	"github.com/nspcc-dev/neo-go/pkg/smartcontract/trigger"
 	"github.com/nspcc-dev/neo-go/pkg/util"
@@ -412,5 +414,74 @@ func TestNewBlockchain_InitHardforks(t *testing.T) {
 			config.HFDomovoi.String():       20,
 			config.HFEchidna.String():       25,
 		}, bc.GetConfig().Hardforks)
+	})
+}
+
+type nopCloserStorage struct {
+	storage.Store
+}
+
+func (nopCloserStorage) Close() error {
+	return nil
+}
+
+func TestBlockchainRestoreStateRootInHeader(t *testing.T) {
+	bc := newTestChainWithCustomCfg(t, func(c *config.Config) {
+		c.ProtocolConfiguration.StateRootInHeader = true
+	})
+	require.NoError(t, bc.AddBlock(bc.newBlock()))
+	require.NoError(t, bc.AddBlock(bc.newBlock()))
+	require.NoError(t, bc.AddBlock(bc.newBlock()))
+
+	w := io.NewBufBinWriter()
+	require.NoError(t, chaindump.Dump(bc, w.BinWriter, 0, 4))
+	require.NoError(t, w.Err)
+
+	data := w.Bytes()
+	fcfg := func(c *config.Config) { c.ProtocolConfiguration.StateRootInHeader = true }
+	initChain := func(st storage.Store) *Blockchain {
+		chain := initTestChain(t, st, fcfg)
+		go chain.Run()
+		return chain
+	}
+
+	t.Run("empty MemoryStore", func(t *testing.T) {
+		bc := initChain(storage.NewMemoryStore())
+		r := io.NewBinReaderFromBuf(data)
+		require.NoError(t, chaindump.Restore(bc, r, 0, 4, nil))
+		bc.Close()
+	})
+	t.Run("not empty MemoryStore, one block", func(t *testing.T) {
+		st := nopCloserStorage{Store: storage.NewMemoryStore()}
+
+		bc := initChain(st)
+		r := io.NewBinReaderFromBuf(data)
+		require.NoError(t, chaindump.Restore(bc, r, 0, 1, nil))
+		expected := bc.stateRoot.CurrentLocalStateRoot()
+		bc.Close()
+
+		bc = initChain(st)
+		actual := bc.stateRoot.CurrentLocalStateRoot()
+		require.Equal(t, expected, actual)
+		r = io.NewBinReaderFromBuf(data)
+		require.NoError(t, chaindump.Restore(bc, r, 1, 1, nil))
+		bc.Close()
+	})
+	t.Run("not empty MemoryStore, multiple blocks", func(t *testing.T) {
+		st := nopCloserStorage{Store: storage.NewMemoryStore()}
+		{
+			bc := initChain(st)
+			r := io.NewBinReaderFromBuf(data)
+			require.NoError(t, chaindump.Restore(bc, r, 0, 2, nil))
+			expected := bc.stateRoot.CurrentLocalStateRoot()
+			bc.Close()
+
+			bc = initChain(st)
+			actual := bc.stateRoot.CurrentLocalStateRoot()
+			require.Equal(t, expected, actual)
+			r = io.NewBinReaderFromBuf(data)
+			require.NoError(t, chaindump.Restore(bc, r, 2, 1, nil))
+			bc.Close()
+		}
 	})
 }

--- a/pkg/core/stateroot/module.go
+++ b/pkg/core/stateroot/module.go
@@ -183,13 +183,15 @@ func (s *Module) Init(height uint32) error {
 		updateStateHeightMetric(h)
 	}
 
-	if height == 0 {
+	if s.mpt == nil {
 		s.mpt = mpt.NewTrie(nil, s.mode, s.Store)
-		s.currentLocal.Store(util.Uint256{})
-		return nil
 	}
 	r, err := s.getStateRoot(makeStateRootKey(height))
 	if err != nil {
+		if height == 0 {
+			s.currentLocal.Store(util.Uint256{})
+			return nil
+		}
 		return err
 	}
 	s.currentLocal.Store(r.Root)

